File: example/js/xpt/serial_send_receive.js
'use strict';

// Portable XPT serial port: open with fixed 8N1 framing, send, receive with a deadline, close.
//
// The device name comes from the command line; without that device the sample reports the
// structured OpenFailed error and returns without failing, which is itself the documented
// no-device behavior. Build with WSE_BUILD_XPT=ON and WSE_BUILD_NODE_BINDING=ON.
//
// XPT deliberately has no default timeout: every operation takes an explicit context so a caller
// can never wait forever by accident. These calls block the calling thread until they finish or
// the deadline elapses, so keep the deadline short on the main thread.
//
//   node example/js/xpt/serial_send_receive.js <installed lang/js directory or wse.node> \
//     <device name> [baud rate]
//
// Argument order: [2] package or addon, [3] device name, [4] optional baud rate.
// Windows uses "COM3"-style names; Linux uses "/dev/ttyUSB0"-style paths.
//
// 8N1 is not a choice this sample makes: eight data bits, no parity, one stop bit is the only
// framing the portable API offers, so the device name and the baud rate are the whole
// configuration. The legacy connector, which does expose framing, is not bound at all.
//
// The first argument is either the installed `lang/js` package directory or a built `wse.node`
// addon. Given the addon the sample still loads the package, so what is exercised is the public
// JavaScript surface an application uses rather than the raw native entry points.

const path = require('node:path');

const target = process.argv[2] || process.env.WSE_JAVASCRIPT_PACKAGE;
if ( !target ) {
  throw new Error('Pass the installed lang/js directory, or the built wse.node addon.');
}
// A `.node` argument is the build-tree addon. Point the package loader at it and still require the
// package, so the sample always demonstrates the same public surface an application uses.
if ( target.endsWith('.node') ) process.env.WSE_NODE_ADDON = path.resolve(target);
const wse = require(target.endsWith('.node')
  ? path.resolve(__dirname, '..', '..', '..', 'lang', 'js')
  : path.resolve(target));

if ( wse.SerialPort === undefined ) {
  throw new Error('This WSE build does not include the XPT component.');
}

// One of the supported rates: 1200/2400/4800/9600/19200/38400/57600/115200.
// The fastest of them is the default because it is the common one for modern USB adapters. A rate
// outside that set is refused at open(); a rate the peer does not share opens successfully and
// then exchanges nothing but noise, which no error can report.
const DEFAULT_BAUD_RATE = 115200;
// The ceiling for one receive, not a length to wait for. Whatever arrived within the deadline is
// returned, up to this many bytes, and anything beyond stays buffered for the next receive.
const RECEIVE_LIMIT_BYTES = 256;

async function main() {
  const deviceName = process.argv[3];
  const baudRate = process.argv[4] === undefined ? DEFAULT_BAUD_RATE : Number(process.argv[4]);
  if ( !deviceName || !Number.isInteger(baudRate) || baudRate <= 0 ) {
    // Deliberate: a device name cannot be guessed and differs per platform, so the sample says
    // what it needs and exits with success rather than failing on a machine with no serial port.
    console.log('Pass a serial device name and an optional baud rate, for example COM3 115200.');
    return;
  }

  // One explicit deadline is reused for every operation in this sample.
  // Reused, not shared: each call gets a fresh 1000 ms, and the object carries no state between
  // them. A second is long for a byte at 115200 baud and short enough that a wrong device name
  // does not leave the sample sitting on the main thread.
  const context = { timeoutMs: 1000 };

  // The object owns the OS handle from open() until close(). JavaScript has no deterministic
  // destructor, so nothing releases that handle on its own: skip close() and the port stays
  // locked against other processes until the garbage collector finalizes this object.
  const port = new wse.SerialPort();
  try {
    port.open(deviceName, baudRate, context);
  } catch (error) {
    // Returning here rather than entering the try/finally below is safe precisely because a
    // failed open leaves nothing open to close.
    // No such device on this machine is a reportable state, not a sample defect.
    console.log(`open ${deviceName} failed:`,
      error.category, error.code, error.nativeCode, error.message);
    console.log('Pass a port that exists on this machine.');
    return;
  }

  try {
    console.log(`opened ${port.deviceName()} at ${port.baudRate()} baud (8N1)`);

    // A serial line carries bytes, not text, so the payload is encoded explicitly. The trailing
    // carriage return is what most devices treat as end of command. send() returns the number of
    // bytes written and keeps writing until the buffer is gone or the deadline expires.
    const payload = Buffer.from('wse-xpt-serial\r', 'utf8');
    console.log('sent bytes:', port.send(payload, context));

    // One receive returns whatever the peer produced within the deadline. Without a loopback plug
    // or an answering device, TimedOut is the expected structured result.
    // That is a result, not a failure: it is checked by code rather than by message text, and the
    // sample carries on to close cleanly and exit with success.
    try {
      const reply = port.receive(RECEIVE_LIMIT_BYTES, context);
      console.log(`received ${reply.length} bytes:`, reply.toString('utf8'));
    } catch (error) {
      if ( error.code === wse.TransportErrorCode.TimedOut ) {
        console.log('no reply within the deadline (expected without an answering device)');
      } else {
        console.log('receive reported:', error.code, error.message);
      }
    }
  } finally {
    // close() ends the session and is idempotent; the object stays usable for a later open().
    // This is the only thing that releases the OS handle at a known moment, which is why it runs
    // in a finally rather than after the last console.log.
    port.close();
  }
  console.log('closed');
}

main().catch((error) => {
  console.error(error.category, error.code, error.nativeCode, error.message);
  process.exitCode = 1;
});

File: example/python/xpt/serial_send_receive.py
"""Portable XPT serial port: open with fixed 8N1 framing, send, receive, close.

The device name and baud rate come from the command line. Without them, or without the physical
device, the sample reports that in one sentence and exits cleanly; the structured OpenFailed error
is itself the documented no-device behavior. Build with WSE_BUILD_XPT=ON and
WSE_BUILD_PYTHON_BINDING=ON.

XPT deliberately has no default timeout: every operation takes an explicit OperationContext so a
caller can never wait forever by accident.

    python example/python/xpt/serial_send_receive.py <path to the built _wse module> <device> <baud>

Windows uses "COM3"-style names; Linux uses "/dev/ttyUSB0"-style paths. The supported rates are
1200/2400/4800/9600/19200/38400/57600/115200.
"""

import importlib.util
import sys
from pathlib import Path

# A second is many character times even at the slowest supported rate, so a device that answers at
# all answers within it. Shortening this would start reporting a slow device as a missing one;
# lengthening it only makes the no-device case, which is the common one here, take longer to report.
TIMEOUT_MS = 1000
# The sample prints what came back, so the limit only has to hold a short echo. It is a ceiling on
# one receive, not a buffer that is always filled: a shorter burst returns as soon as it arrives.
RECEIVE_LIMIT = 256


def load_wse():
    """Loads the binding, either as the installed package or from a built module path.

    With no argument the installed `wse` package is imported. With one, the built `_wse` module at
    that path is loaded directly, which is how the sample runs against a build tree. main() below
    requires the path, so only the second mode is reachable here.
    """
    if len(sys.argv) == 1:
        import wse

        return wse
    module_path = Path(sys.argv[1]).resolve()
    spec = importlib.util.spec_from_file_location("_wse", module_path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"Unable to load WSE Python module: {module_path}")
    module = importlib.util.module_from_spec(spec)
    sys.modules[spec.name] = module
    spec.loader.exec_module(module)
    return module


def report_transport_error(operation: str, failure) -> None:
    """Prints one structured line for a WseError: the operation, then the whole error identity.

    Every field is printed because they answer different questions. `category` says which layer
    refused, `code` is the portable TransportErrorCode a caller may branch on, `native_code` is the
    untranslated errno or Win32 status a platform-specific diagnosis needs, and the message is
    prose for a human. A caller branches on category and code, never on the message text.
    """
    print(
        f"ERROR: {operation} failed. category={failure.category} code={failure.code}"
        f" native={failure.native_code} message={failure}"
    )


def main() -> int:
    # Argument handling first, before the binding is loaded: a usage mistake should not depend on
    # whether the native module is present. Every bad-argument path prints and returns 0, because
    # running the sample wrongly is not a build failure.
    if len(sys.argv) < 4:
        print(
            "No serial device was given; pass the built _wse module path, then a device name and a"
            " baud rate: python example/python/xpt/serial_send_receive.py <module path>"
            " <device> <baud>"
        )
        return 0
    device_name = sys.argv[2]
    try:
        baud_rate = int(sys.argv[3])
    except ValueError:
        print(f"The baud rate argument is not a number: {sys.argv[3]}")
        return 0

    wse = load_wse()
    # A build without WSE_BUILD_XPT registers no SerialPort at all, so the name has to be looked
    # for rather than caught; there is no stub that raises a friendlier error later.
    if not hasattr(wse, "SerialPort"):
        raise SystemExit("This WSE build does not include the XPT component.")

    # One explicit deadline is reused for every operation in this sample.
    # A context is a plain value and carries no port state, so sharing one across open, send, and
    # receive is safe; each call starts its own countdown from that same duration rather than
    # spending one shared budget.
    context = wse.OperationContext(TIMEOUT_MS)

    # The context manager is the single owner; leaving it closes the port.
    # That matters more here than for a socket: a serial device is usually exclusive, so a port
    # left open by a crashed sample keeps every other program off it until the process dies.
    with wse.SerialPort() as port:
        try:
            port.open(device_name, baud_rate, context)
        except wse.WseError as failure:
            # No device on this machine is a reportable state, not a sample defect.
            # OPEN_FAILED covers a name that does not exist and a port another program already
            # holds; CONFIGURATION_FAILED covers a device that exists but refuses the rate.
            # Returning 0 is deliberate so the sample runs on a machine with no serial hardware.
            report_transport_error(f"open {device_name}", failure)
            print("point the device argument at a port that exists on this machine")
            return 0
        # Read back rather than echo the arguments: these are what the port actually settled on.
        # The framing is fixed at 8 data bits, no parity, one stop bit, so only the rate is a
        # choice; a device expecting anything else cannot be talked to through this API.
        print(f"opened {port.device_name} at {port.baud_rate} baud (8N1)")

        # send() writes the whole buffer or fails on the deadline, so the returned count equals the
        # payload length on success. The trailing carriage return is the line terminator most simple
        # serial devices expect before they treat a command as complete.
        print("sent bytes:", port.send(b"WSE-XPT-SERIAL\r", context))

        # One receive returns whatever the peer produced within the deadline. Without a loopback
        # plug or an answering device, a timeout is the expected structured result.
        try:
            payload = port.receive(RECEIVE_LIMIT, context)
            print(f"received {len(payload)} bytes:", payload.decode("utf-8", "replace"))
        except wse.WseError as failure:
            if failure.code == wse.TransportErrorCode.TIMED_OUT.value:
                print("no reply within the deadline (expected without an answering device)")
            else:
                report_transport_error("receive", failure)
                return 1

        # close() is idempotent; leaving the context manager would do the same.
        port.close()
        print("closed")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
